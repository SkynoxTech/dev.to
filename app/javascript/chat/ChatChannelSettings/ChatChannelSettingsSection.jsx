import { h } from 'preact';
import PropTypes from 'prop-types';

import ModSection from './ModSection';
import PersonalSettings from './PersonalSetting';
import LeaveMembershipSection from './LeaveMembershipSection';
import ModFaqSection from './ModFaqSection';
import ChannelDescriptionSection from './ChannelDescriptionSection';
import ChatChannelMembershipSection from './ChatChannelMembershipSection';

const ChatChannelSettingsSection = ({
  channelDiscoverable,
  updateCurrentMembershipNotificationSettings,
  handleleaveChatChannelMembership,
  handlePersonChatChennelSetting,
  handleChannelDescriptionChanges,
  handleChannelDiscoverableStatus,
  handleDescriptionChange,
  handleChatChannelInvitations,
  handleInvitationUsernames,
  toggleScreens,
  removeMembership,
  chatChannelAcceptMembership,
  channelDescription,
  chatChannel,
  currentMembership,
  activeMemberships,
  pendingMemberships,
  requestedMemberships,
  invitationUsernames,
  showGlobalBadgeNotification,
}) => (
  <div>
    <ChannelDescriptionSection
      channelName={chatChannel.name}
      channelDescription={chatChannel.description}
      currentMembershipRole={currentMembership.role}
      className="channel-description-section"
    />
    <ChatChannelMembershipSection
      currentMembershipRole={currentMembership.role}
      activeMemberships={activeMemberships}
      removeMembership={removeMembership}
      pendingMemberships={pendingMemberships}
      requestedMemberships={requestedMemberships}
      chatChannelAcceptMembership={chatChannelAcceptMembership}
      toggleScreens={toggleScreens}
      className="channel-membership-sections"
    />
    <ModSection
      invitationUsernames={invitationUsernames}
      handleInvitationUsernames={handleInvitationUsernames}
      handleChatChannelInvitations={handleChatChannelInvitations}
      channelDescription={channelDescription}
      handleDescriptionChange={handleDescriptionChange}
      channelDiscoverable={channelDiscoverable}
      handleChannelDiscoverableStatus={handleChannelDiscoverableStatus}
      handleChannelDescriptionChanges={handleChannelDescriptionChanges}
      currentMembershipRole={currentMembership.role}
      className="channel-mod-section"
    />
    <PersonalSettings
      updateCurrentMembershipNotificationSettings={
        updateCurrentMembershipNotificationSettings
      }
      showGlobalBadgeNotification={showGlobalBadgeNotification}
      handlePersonChatChennelSetting={handlePersonChatChennelSetting}
      className="channel-personal-seeting"
    />
    <LeaveMembershipSection
      currentMembershipRole={currentMembership.role}
      handleleaveChatChannelMembership={handleleaveChatChannelMembership}
      className="channel-leave-membership-section"
    />
    <ModFaqSection
      currentMembershipRole={currentMembership.role}
      email="yo@dev.to"
      className="channel-mod-faq"
    />
  </div>
);

ChatChannelSettingsSection.propTypes = {
  chatChannel: PropTypes.isRequired,
  currentMembership: PropTypes.isRequired,
  activeMemberships: PropTypes.isRequired,
  pendingMemberships: PropTypes.isRequired,
  requestedMemberships: PropTypes.isRequired,
  invitationUsernames: PropTypes.string.isRequired,
  channelDescription: PropTypes.string.isRequired,
  channelDiscoverable: PropTypes.bool.isRequired,
  showGlobalBadgeNotification: PropTypes.bool.isRequired,
  handleleaveChatChannelMembership: PropTypes.func.isRequired,
  chatChannelAcceptMembership: PropTypes.func.isRequired,
  removeMembership: PropTypes.func.isRequired,
  toggleScreens: PropTypes.func.isRequired,
  handleInvitationUsernames: PropTypes.func.isRequired,
  handleChatChannelInvitations: PropTypes.func.isRequired,
  handleDescriptionChange: PropTypes.func.isRequired,
  handleChannelDiscoverableStatus: PropTypes.func.isRequired,
  handleChannelDescriptionChanges: PropTypes.func.isRequired,
  handlePersonChatChennelSetting: PropTypes.func.isRequired,
  updateCurrentMembershipNotificationSettings: PropTypes.func.isRequired,
};

export default ChatChannelSettingsSection;
