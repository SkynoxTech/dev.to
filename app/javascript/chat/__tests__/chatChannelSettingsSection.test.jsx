import { h } from 'preact';
import render from 'preact-render-to-json';
import { shallow } from 'preact-render-spy';
import ChatChannelSettingsSection from '../ChatChannelSettings/ChatChannelSettingsSection';

const propData = {
  channelDiscoverable: true,
  chatChannel: {
    name: 'dummy-channel',
    description: 'dummy-description',
    id: 1,
  },
  currentMembership: {
    role: 'mod',
    user_id: 1,
    chat_channel_id: 1,
    id: 2,
    status: 'active',
  },
  pendingMemberships: [
    {
      name: 'test user',
      username: 'testusername',
      user_id: '1',
      membership_id: '2',
      role: 'member',
      status: 'pending',
      image: '',
    },
  ],
  requestedMemberships: [
    {
      name: 'test user',
      username: 'testusername',
      user_id: '1',
      membership_id: '2',
      role: 'member',
      status: 'requested',
      image: '',
    },
  ],
  activeMemberships: [
    {
      name: 'test user',
      username: 'testusername',
      user_id: '1',
      membership_id: '2',
      role: 'mod',
      status: 'active',
      image: '',
    },
  ],
};

const getChatChannelSettingsSection = (data) => (
  <ChatChannelSettingsSection
    channelDiscoverable={data.channelDiscoverable}
    chatChannel={data.chatChannel}
    currentMembership={data.currentMembership}
    pendingMemberships={data.pendingMemberships}
    requestedMemberships={data.requestedMemberships}
    activeMemberships={data.activeMemberships}
  />
);

describe('<ChatChannelSettingsSection />', () => {
  it('should render & test snapshot', () => {
    const tree = render(getChatChannelSettingsSection(propData));
    expect(tree).toMatchSnapshot();
  });

  it('should have the chat channel description component', () => {
    const context = shallow(getChatChannelSettingsSection(propData));
    expect(context.find('.channel-description-section').exists()).toEqual(true);
  });

  it('shoul render the chat channel membership section', () => {
    const context = shallow(getChatChannelSettingsSection(propData));
    expect(context.find('.channel-membership-sections').exists()).toEqual(true);
  });

  it('shoul render the chat channel mod section', () => {
    const context = shallow(getChatChannelSettingsSection(propData));
    expect(context.find('.channel-mod-section').exists()).toEqual(true);
  });

  it('shoul render the chat channel personal section', () => {
    const context = shallow(getChatChannelSettingsSection(propData));
    expect(context.find('.channel-personal-seeting').exists()).toEqual(true);
  });

  it('shoul render the chat channel leave membership section', () => {
    const context = shallow(getChatChannelSettingsSection(propData));
    expect(context.find('.channel-leave-membership-section').exists()).toEqual(
      true,
    );
  });

  it('shoul render the chat channel mod faq section', () => {
    const context = shallow(getChatChannelSettingsSection(propData));
    expect(context.find('.channel-mod-faq').exists()).toEqual(true);
  });
});
