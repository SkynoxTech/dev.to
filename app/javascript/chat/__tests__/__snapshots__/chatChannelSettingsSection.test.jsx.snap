// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatChannelSettingsSection /> should render & test snapshot 1`] = `
<div>
  <div
    class="p-4 grid gap-2 crayons-card mb-4 channel_details"
  >
    <h1
      class="mb-1 channel_title"
    >
      dummy-channel
    </h1>
    <p>
      dummy-description
    </p>
    <p
      class="fw-bold"
    >
      You are a channel mod
    </p>
  </div>
  <div
    class="membership-list"
  >
    <div
      class="p-4 grid gap-2 crayons-card mb-4"
    >
      <h3
        class="mb-2 active_members"
      >
        Members
      </h3>
      <div
        class="flex items-center"
      >
        <a
          class="chatmessagebody__username--link"
          data-content="sidecar-user"
          href="/testusername"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="crayons-avatar crayons-avatar--l mr-3"
          >
            <img
              alt="test user profile"
              class="crayons-avatar__image align-middle"
              role="presentation"
              src=""
            />
          </span>
          <span
            class="mr-2 user_name"
          >
            test user
          </span>
        </a>
      </div>
      <div
        class="row view-membership-btn"
      >
        <button
          class="crayons-btn align-center crayons-btn--s view-all-memberships"
          onClick={[Function]}
          type="button"
        >
          View All
        </button>
      </div>
    </div>
    <div
      class="p-4 grid gap-2 crayons-card mb-4 pending_memberships"
    >
      <h3
        class="mb-2"
      >
        Pending Invitations
      </h3>
      <div
        class="flex items-center"
      >
        <a
          class="chatmessagebody__username--link"
          data-content="sidecar-user"
          href="/testusername"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="crayons-avatar crayons-avatar--l mr-3"
          >
            <img
              alt="test user profile"
              class="crayons-avatar__image align-middle"
              role="presentation"
              src=""
            />
          </span>
          <span
            class="mr-2 user_name"
          >
            test user
          </span>
        </a>
        <button
          class="crayons-btn crayons-btn--icon-rounded crayons-btn--ghost remove-membership"
          data-membership-id="2"
          data-membership-status="pending"
          type="button"
        >
          x
        </button>
      </div>
    </div>
    <div
      class="p-4 grid gap-2 crayons-card mb-4"
    >
      <h3
        class="mb-2 requested_memberships"
      >
        Joining Request
      </h3>
      <div
        class="flex items-center"
      >
        <a
          class="chatmessagebody__username--link"
          data-content="sidecar-user"
          href="/testusername"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            class="crayons-avatar crayons-avatar--l mr-3"
          >
            <img
              alt="test user profile"
              class="crayons-avatar__image align-middle"
              role="presentation"
              src=""
            />
          </span>
          <span
            class="mr-2 user_name"
          >
            test user
          </span>
        </a>
        <button
          class="crayons-btn crayons-btn--icon-rounded crayons-btn--ghost add-membership"
          data-membership-id="2"
          type="button"
        >
          +
        </button>
        <button
          class="crayons-btn crayons-btn--icon-rounded crayons-btn--ghost remove-membership"
          data-membership-id="2"
          data-membership-status="requested"
          type="button"
        >
          x
        </button>
      </div>
    </div>
  </div>
  <div
    class="mod-section"
  >
    <div
      class="crayons-card p-4 grid gap-2 mb-4"
    >
      <div
        class="crayons-field"
      >
        <label
          class="crayons-field__label invitation_form_title"
          htmlFor="chat_channel_membership_Usernames to Invite"
        >
          Usernames to invite
        </label>
        <input
          class="crayons-textfield"
          id="chat_channel_membership_invitation_usernames"
          name="chat_channel_membership[invitation_usernames]"
          placeholder="Comma separated"
          type="text"
        />
      </div>
      <div>
        <button
          class="crayons-btn"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
    <div
      class="crayons-card p-4 grid gap-2 mb-4 settings-section"
    >
      <h3>
        Channel Settings
      </h3>
      <div
        class="crayons-field"
      >
        <label
          class="crayons-field__label"
          htmlFor="chat_channel_description"
        >
          Description
        </label>
        <textarea
          class="crayons-textfield"
          id="chat_channel_description"
          name="description"
        />
      </div>
      <div
        class="crayons-field crayons-field--checkbox"
      >
        <input
          checked={true}
          class="crayons-checkbox"
          id="c2"
          type="checkbox"
        />
        <label
          class="crayons-field__label"
          htmlFor="c2"
        >
          Channel Discoverable
        </label>
      </div>
      <div>
        <button
          class="crayons-btn"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
  <div
    class="crayons-card p-4 grid gap-2 mb-4 personl-settings"
  >
    <h3>
      Personal Settings
    </h3>
    <h4>
      Notifications
    </h4>
    <div
      class="crayons-field crayons-field--checkbox"
    >
      <input
        class="crayons-checkbox"
        id="c3"
        type="checkbox"
      />
      <label
        class="crayons-field__label"
        htmlFor="c3"
      >
        Receive Notifications for New Messages
      </label>
    </div>
    <div>
      <button
        class="crayons-btn"
        type="submit"
      >
        Submit
      </button>
    </div>
  </div>
  <div
    class="crayons-card grid gap-2 p-4 faq-section"
  >
    <p
      class="contact-details"
    >
      Questions about Connect Channel moderation? Contact
      <a
        class="mx-2 url-link"
        href="mailto:yo@dev.to"
        rel="noopener noreferrer"
        target="_blank"
      >
        yo@dev.to
      </a>
    </p>
  </div>
</div>
`;
